### DISCUSSION TITLE
Review :  p2p: Fill reconciliation sets (Erlay)

### DISCUSSION THREAD

**Discussion by Prabhat1308** [Bitshala/BitcoinCore-PR-Review-Club#68] [PR Reviews]
# Session Details
Date : 23-05-2024
Time: IST 20:00 (UTC 14:30)
Link : [p2p: Fill reconciliation sets (Erlay)](https://github.com/bitcoin/bitcoin/pull/30116)
BIP : 330 
Difficulty : High

## Notes
This PR introduces tracking of per-peer reconciliation sets containing transactions to be exchanged efficiently using minisketch.
Currently all the transaction are flooded to each peer via flooding algorithm .

## Questions
1. What is the objective of the PR ?
2. What are benchmark functions ?
3. Can you decipher how is the combined salt being formed ? Any reason that you can think of ?
4. We form a local salt when Pre-registering a peer . Why pre-register in the first place ? Any pros you can think of ?
5. Why testing with Boost library and not use python tests ?

## Learnings 
sketches
erlay
multi-threaded architecture

## Pre-Requisites
PR review club - BIP330

**Comment by YeamlaksiraKebede** [comment]
Answers for the questions
1. I feel like this PR tries to answer some potential points of misunderstanding, like, latency increase, bandwidth savings, security implications and compatibility with existing protocols.
2) Benchmark functions are routines used to measure and evaluate the performance of software, algorithms, or systems. They provide insights into how well a piece of code performs under various conditions, which can be crucial for optimization and ensuring that performance requirements are met. In the context of Bitcoin and the Erlay BIP, benchmark functions might be used to: Transaction Propagation, Comparison of Protocols, Scalability, Resource Efficiency.
3) In the Erlay Bitcoin Improvement Proposal (BIP), the term "combined salt" likely refers to the process of creating a single value from two separate salts contributed by different parties. Let's break down how this combined salt is formed and consider the reasons behind this approach:
A. Contributed Salts:
   - Each party involved in the transaction relay process contributes a salt, denoted as "salt1" and "salt2".
   - These salts serve as sources of entropy, adding randomness to the computation process.

B. Sorting Salts:
   - The two salts are sorted in ascending order to ensure consistency in their representation.
   - Sorting ensures that the same combined salt is generated regardless of the order in which the salts were contributed.

C. Hash Computation:
   - The sorted salts are then combined and hashed using a cryptographic hash function, such as the TaggedHash function specified by BIP-340.
   - Hashing combines the entropy from both salts to create a single value that is difficult to reverse-engineer.

D. Reasons for Combined Salt:
   - Increased Entropy: By combining two separate salts, the resulting combined salt contains more entropy than either salt individually.
   - Uniformity: Sorting the salts ensures that the same combined salt is generated regardless of the order of contribution, promoting consistency in the computation process.
   - Security: Hashing the combined salt helps protect the integrity of the transaction relay process by obscuring the original salts and adding cryptographic security to the computation.
Summary:
The combined salt in the Erlay Bitcoin Improvement Proposal is formed by sorting and hashing two separate salts contributed by different parties. This approach increases entropy, promotes consistency, and enhances security in the transaction relay process.

4) In the context of the Erlay Bitcoin Improvement Proposal (BIP), pre-registering a peer involves forming a local salt in anticipation of future interactions with that peer. Let's explore the reasons why pre-registering peers might be beneficial and some potential advantages:
A. Efficient Communication:
   - Pre-registering a peer allows a node to establish initial communication parameters and prepare for efficient data exchange when interacting with that peer in the future.
   - By pre-registering, nodes can streamline the process of exchanging transaction information and minimize the overhead associated with establishing connections on the fly.

B. Optimized Resource Usage:
   - Pre-registering peers enables nodes to allocate resources more efficiently, such as network bandwidth and processing power, by anticipating upcoming interactions and preparing accordingly.
   - This proactive approach helps nodes manage their resources more effectively, especially in networks with a large number of peers.

C. Enhanced Security:
   - Pre-registering peers may enhance security by allowing nodes to establish trust relationships and verify the authenticity of incoming data more effectively.
   - By pre-registering trusted peers, nodes can prioritize interactions with known entities and mitigate the risk of malicious actors or network attacks.

D. Reduced Latency:
   - Pre-registering peers can potentially reduce latency in data transmission by minimizing the time required to establish connections and negotiate communication parameters.
   - This can lead to faster transaction relay and improved overall network performance.

E. Customized Interaction:
   - Pre-registering peers enables nodes to customize their interaction strategies based on the characteristics and behavior of specific peers.
   - Nodes can adapt their communication protocols and data exchange mechanisms to suit the preferences and requirements of individual peers, optimizing overall network efficiency.

Summary:
Pre-registering peers in the Erlay BIP allows nodes to anticipate future interactions, streamline communication, optimize resource usage, enhance security, reduce latency, and customize interaction strategies. These benefits contribute to improved efficiency, performance, and reliability in the Bitcoin network.

5) Testing with the Boost library rather than using Python tests may offer certain advantages in specific contexts, particularly when developing software in C++ like the Erlay Bitcoin Improvement Proposal (BIP). Here are some reasons why testing with Boost might be preferred:

A. Native Integration:
   - Boost is a widely used library for C++ development and provides native support for writing unit tests.
   - Since the Erlay BIP likely involves C++ code, using Boost for testing allows developers to seamlessly integrate tests into their existing codebase without introducing additional dependencies.

B. Performance:
   - Testing with Boost may offer better performance characteristics compared to using Python tests, especially for low-level code or performance-critical components.
   - Boost's testing framework is designed to be efficient and lightweight, making it suitable for testing complex algorithms or data structures.

C. Language Compatibility:
   - Boost is written in C++ and is specifically designed to work with C++ code.
   - Testing with Boost allows developers to leverage the full power of the C++ language and its features, which may be necessary for testing certain aspects of the Erlay implementation.

D. Robustness and Reliability:
   - Boost is a mature and well-tested library that has been used in numerous production projects.
   - By using Boost for testing, developers can benefit from its robustness, reliability, and extensive feature set, ensuring thorough and effective testing of the Erlay implementation.

E. Community Support:
   - Boost has a large and active community of developers who contribute to its development and maintenance.
   - Testing with Boost allows developers to tap into this community for support, documentation, and resources, which can be valuable when developing complex software like the Erlay BIP.

While using Python tests may also be a viable option, particularly for higher-level or integration testing, testing with Boost offers specific advantages for testing C++ code, such as performance, language compatibility, and native integration. Ultimately, the choice of testing framework depends on the specific requirements and constraints of the project, as well as the preferences of the development team.
